pytorch/pytorch_ddp_nompi.py

```diff
@@ -0,0 +1,242 @@
+            loss.backward()
+            optimizer.step()
+
+        # Validation of the model.
+        model.eval()
+        correct = 0
+        with torch.no_grad():
+            for data, target in valid_loader:
+                data, target = data.view(data.size(0), -1).to(device_id), target.to(device_id)
+                output = model(data)
+                # Get the index of the max log-probability.
+                pred = output.argmax(dim=1, keepdim=True)
+                correct += pred.eq(target.view_as(pred)).sum().item()
+
+        correct_tensor = torch.tensor([correct], dtype=torch.int).to(device_id)
+        dist.all_reduce(correct_tensor)
+        total_correct = correct_tensor.item()
+        accuracy = total_correct / len(valid_loader.dataset)
+
+        trial.report(accuracy, epoch)
+
+        # Handle pruning based on the intermediate value.
+        if trial.should_prune():
+            raise optuna.exceptions.TrialPruned(f"Trial was pruned at epoch {epoch}.")
+
+    return accuracy
+
+
+def setup(backend, rank, world_size):
+    os.environ["MASTER_ADDR"] = "localhost"
+    os.environ["MASTER_PORT"] = "12355"
+    dist.init_process_group(backend, rank=rank, world_size=world_size)
+    print(f"Using {backend} backend.")
+
+
+def cleanup():
+    dist.destroy_process_group()
+
+
+def run_optimize(rank, world_size, device_ids, return_dict):
+    devi = device_ids if device_ids == "cpu" else device_ids[rank]
+    print(f"Running basic DDP example on rank {rank} device {devi}.")
+
+    # Set environmental variables required by torch.distributed.
+    backend = "gloo"
+    if torch.distributed.is_nccl_available():
+        if devi != "cpu":
+            backend = "nccl"
+    setup(backend, rank, world_size)
+
+    if rank == 0:
+        study = optuna.create_study(direction="maximize")
+        study.optimize(
+            partial(objective, device_id=devi),
+            n_trials=N_TRIALS,
+            timeout=300,
+        )
+        return_dict["study"] = study
+    else:
+        for _ in range(N_TRIALS):
+            try:
+                objective(None, devi)
+            except optuna.TrialPruned:
+                pass
+
+    cleanup()
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="PyTorch distributed data-parallel training example."
+    )
+    parser.add_argument(
+        "--device_ids",
+        "-d",
+        nargs="+",
+        type=int,
+        default=[0],
+        help="Specify device_ids if using GPUs.",
+    )
+    parser.add_argument(
+        "--no-cuda", action="store_true", default=False, help="Disables CUDA training."
+    )
+    args = parser.parse_args()
+    if args.no_cuda:
+        device_ids = "cpu"
+    else:
+        device_ids = args.device_ids
+
+    # Download dataset before starting the optimization.
+    datasets.FashionMNIST(DIR, train=True, download=True)
+
+    world_size = len(device_ids) if isinstance(device_ids, list) else 1
```

Comment:
Similarly to [@nzw0301 's comment](https://github.com/optuna/optuna-examples/pull/155#discussion_r1092200101), `device_ids` can take two different types of values, i.e., `str` and `List[int]`. How about accepting only `List[int]` ?

```suggestion
    if args.no_cuda:
        device_ids = []
    else:
        device_ids = args.device_ids

    # Download dataset before starting the optimization.
    datasets.FashionMNIST(DIR, train=True, download=True)

    # device_ids can be empty, but at least a single process is required.
    world_size = max(len(device_ids), 1)
```

Reply:
Thank you fro the suggestion. It's now incorporated in the latest update.